File: tts.js
/* tts.js - compact front end for klatt80.js: phones -> frames. MIT. */
"use strict";
const VOWELS = {IY:[270,2290,3010,60,90,150],IH:[390,1990,2550,65,100,160],EH:[530,1840,2480,70,100,160],
AE:[660,1720,2410,80,110,170],AA:[730,1090,2440,85,110,170],AO:[570,840,2410,80,110,170],
UH:[440,1020,2240,70,100,160],UW:[300,870,2240,60,95,150],AH:[640,1190,2390,80,110,170],
AX:[500,1500,2500,75,110,170],ER:[490,1350,1690,70,100,140]};
const DIPH = {EY:['EH','IY'],AY:['AA','IY'],OY:['AO','IY'],OW:['AX','UH'],AW:['AA','UH']};
const STOPS = {P:[70,{af:.75,ab:.45},35,.55,0],T:[65,{af:.85,a4:.55,a5:.45},35,.60,0],K:[70,{af:.80,a3:.65},35,.55,0],
B:[55,{af:.45,ab:.30},12,.20,1],D:[50,{af:.55,a4:.40,a5:.30},12,.20,1],G:[55,{af:.50,a3:.45},12,.20,1]};
const FRICS = {S:[115,{af:.85,a5:.55,a6:.95,f6:4650,b6:380}],Z:[90,{af:.60,a5:.45,a6:.75,f6:4650,b6:380,av:.42}],
SH:[110,{af:.80,a3:.30,a4:.80,a5:.35,f4:2600,b4:350}],ZH:[85,{af:.55,a3:.25,a4:.60,a5:.30,f4:2600,b4:350,av:.42}],
F:[95,{af:.40,ab:.50}],V:[75,{af:.28,ab:.38,av:.48}],TH:[85,{af:.32,ab:.42}],DH:[70,{af:.22,ab:.32,av:.50}],HH:[70,{ah:.60}]};
const NASALS = {M:[95,1000,1000],N:[90,1700,1900],NG:[100,2600,2900]};
const LIQUIDS = {L:[85,380,1350,2600,130,220],R:[85,460,1380,1720,110,180]};
const GLIDES = {W:[65,300,610,2200],Y:[65,270,2290,3010]};
const AFFR = {CH:['T','SH'],JH:['D','ZH']};
const LTS = {a:'AH',b:'B',c:'K',d:'D',e:'EH',f:'F',g:'G',h:'HH',i:'IH',j:'JH',k:'K',l:'L',m:'M',n:'N',
o:'AA',p:'P',q:'K',r:'R',s:'S',t:'T',u:'AH',v:'V',w:'W',x:'K S',y:'Y',z:'Z'};
/* small built-in lexicon (narration vocabulary + common words); ARPAbet with stress */
const DICT = {
it:'IH1 T',begins:'B IH0 G IH1 N Z',with:'W IH1 DH',one:'W AH1 N',of:'AH1 V',you:'Y UW1',us:'AH1 S',
made:'M EY1 D',pictures:'P IH1 K CH ER0 Z',the:'DH AH0',world:'W ER1 L D',all:'AO1 L',looked:'L UH1 K T',
up:'AH1 P',and:'AH0 N D',still:'S T IH1 L',was:'W AH1 Z',humanico:'HH UW0 M AA0 N IY1 K OW0',for:'F AO1 R',
humanity:'HH Y UW0 M AE1 N AH0 T IY0',i:'AY1',did:'D IH1 D',not:'N AA1 T',build:'B IH1 L D',any:'EH1 N IY0',
this:'DH IH1 S',cathedrals:'K AH0 TH IY1 D R AH0 L Z',wire:'W AY1 ER0',raised:'R EY1 Z D',by:'B AY1',hand:'HH AE1 N D',
your:'Y AO1 R',hands:'HH AE1 N D Z',everyone:'EH1 V R IY0 W AH2 N',can:'K AE1 N',do:'D UW1',again:'AH0 G EH1 N',
people:'P IY1 P AH0 L',who:'HH UW1',a:'AH0',voice:'V OY1 S',in:'IH1 N',every:'EH1 V R IY0',room:'R UW1 M',
technology:'T EH0 K N AA1 L AH0 JH IY0',is:'IH1 Z',them:'DH EH1 M',am:'AE1 M',only:'OW1 N L IY0',are:'AA1 R',
always:'AO1 L W EY0 Z',built:'B IH1 L T',bridges:'B R IH1 JH AH0 Z',laid:'L EY1 D',grids:'G R IH1 D Z',
cities:'S IH1 T IY0 Z',touched:'T AH1 CH T',sky:'S K AY1',whatever:'W AH0 T EH1 V ER0',comes:'K AH1 M Z',
next:'N EH1 K S T',make:'M EY1 K',beautiful:'B Y UW1 T AH0 F AH0 L',we:'W IY1',to:'T UW1',too:'T UW1',
love:'L AH1 V',bots:'B AA1 T S',draw:'D R AO1',perfect:'P ER1 F AH0 K T',lines:'L AY1 N Z',wobble:'W AA1 B AH0 L',
beautifully:'B Y UW1 T AH0 F AH0 L IY0',people2:'P IY1 P AH0 L',hello:'HH AH0 L OW1',hi:'HH AY1',
yes:'Y EH1 S',no:'N OW1',speech:'S P IY1 CH',synth:'S IH1 N TH',machine:'M AH0 SH IY1 N',human:'HH Y UW1 M AH0 N'};
function wordPhones(w){
  w = w.toLowerCase().replace(/[^a-z']/g,'');
  if (DICT[w]) return DICT[w].split(' ').map(p=>{const m=p.match(/^([A-Z]+)([012])?$/);return [m[1], m[2]?+m[2]:0];});
  const out=[]; for (const ch of w) if (LTS[ch]) for (const p of LTS[ch].split(' ')) out.push([p,-1]);
  return out;
}
function vowelTgt(v, av){ const [f1,f2,f3,b1,b2,b3]=VOWELS[v]; return {av,f1,f2,f3,b1,b2,b3}; }
function segments(phseq){
  const segs=[]; const n=phseq.length;
  for (let i=0;i<n;i++){
    const [ph,stress]=phseq[i]; const nxt=i+1<n?phseq[i+1][0]:null; const stressed=stress===1;
    if (VOWELS[ph]) segs.push({t:vowelTgt(ph,.92),dur:stressed?150:(stress===2?100:85),tr:50,kind:'vowel',stress});
    else if (DIPH[ph]){ const [a,b]=DIPH[ph]; const d=stressed?165:115;
      segs.push({t:vowelTgt(a,.92),dur:d*.45|0,tr:50,kind:'vowel',stress});
      segs.push({t:vowelTgt(b,.92),dur:d*.55|0,tr:d*.5|0,kind:'vowel',stress}); }
    else if (STOPS[ph]){ const [cl,burst,asp,ahv,voiced]=STOPS[ph];
      const clo = voiced?{av:.18,f1:200,b1:250,f2:1000,f3:2200}:{};
      segs.push({t:clo,dur:cl,tr:3,kind:'sil'}); segs.push({t:Object.assign({},burst),dur:14,tr:3,kind:'burst'});
      if (asp>0){ const ah={ah:ahv}; if (nxt&&VOWELS[nxt]){const v=vowelTgt(nxt);ah.f1=v.f1;ah.f2=v.f2;ah.f3=v.f3;}
        segs.push({t:ah,dur:asp,tr:12,kind:'aspir'}); } }
    else if (AFFR[ph]){ const [st,fr]=AFFR[ph];
      const sub=segments([[st,-1],[fr,-1]]).filter(s=>s.kind!=='aspir'); sub[0].dur=sub[0].dur*.7|0; segs.push(...sub); }
    else if (FRICS[ph]){ const [dur,prm]=FRICS[ph]; const t=Object.assign({},prm);
      if (ph==='HH'&&nxt&&VOWELS[nxt]){const v=vowelTgt(nxt);t.f1=v.f1;t.f2=v.f2;t.f3=v.f3;}
      segs.push({t,dur,tr:20,kind:'fric'}); }
    else if (NASALS[ph]){ const [dur,f2,fnz]=NASALS[ph];
      segs.push({t:{av:.75,f1:270,b1:250,f2,b2:200,f3:2400,b3:250,fnz,bnz:250},dur,tr:35,kind:'nasal'}); }
    else if (LIQUIDS[ph]){ const [dur,f1,f2,f3,b1,b2]=LIQUIDS[ph];
      segs.push({t:{av:.72,f1,f2,f3,b1,b2,b3:200},dur,tr:45,kind:'liquid'}); }
    else if (GLIDES[ph]){ const [dur,f1,f2,f3]=GLIDES[ph];
      segs.push({t:{av:.80,f1,f2,f3,b1:90,b2:130,b3:180},dur,tr:45,kind:'glide'}); }
  }
  return segs;
}
/* calibration: reuse Python-computed source scales per kind via steady-state unit RMS
   (computed once here at load, same algorithm as Python). */
const LEVEL = {vowel:0,nasal:-4,liquid:-4,glide:-2,fric:-7,fric_weak:-10,burst:-5,aspir:-12,voicebar:-20,fric_voice:-16};
const _rc = {};
function unitRMS(params, source){
  const key = source+JSON.stringify(params);
  if (_rc[key]) return _rc[key];
  const v = new Voice(); Object.assign(v, params);
  v.av=0;v.ah=0;v.af=0; v[source]=1; if(!params.f0)v.f0=110;
  let acc=0,cnt=0;
  for(let f=0;f<70;f++){ const out=v.frame(50); if(f>=30) for(let i=0;i<50;i++){acc+=out[i]*out[i];cnt++;} }
  return _rc[key]=Math.sqrt(acc/cnt)+1e-9;
}
let _vref=null;
function vref(){ if(!_vref)_vref=unitRMS({f1:730,b1:85,f2:1090,b2:110,f3:2440,b3:170},'av'); return _vref; }
function balance(t, kind){
  t=Object.assign({},t);
  const set=(src,k)=>{const want=vref()*Math.pow(10,LEVEL[k]/20);t[src]=want/unitRMS(t,src);};
  if(kind==='vowel')set('av','vowel');
  else if(kind==='nasal'||kind==='liquid')set('av',kind);
  else if(kind==='glide')set('av','glide');
  else if(kind==='fric'){set('af',(t.ab&&!t.a4&&!t.a5&&!t.a6)?'fric_weak':'fric');if(t.av)set('av','fric_voice');}
  else if(kind==='burst')set('af','burst');
  else if(kind==='aspir')set('ah','aspir');
  else if(kind==='voicebar')set('av','voicebar');
  return t;
}
function framesForText(text, rate=0.85, pitch=1.0){
  const tokens = text.replace(/[\u2014\u2013]/g,' - ').match(/[A-Za-z']+|[.,!?;:\-]/g)||[];
  const words=[]; 
  for(const tk of tokens){
    if(/^[.,!?;:\-]$/.test(tk)){ if(words.length){words[words.length-1].pause={'.':340,'!':340,'?':340,',':200,';':220,':':220,'-':160}[tk]||200; if('.!?'.includes(tk))words.push({sentenceEnd:true});} }
    else {const ph=wordPhones(tk); if(ph.length)words.push({w:tk,ph,pause:0});}
  }
  const frames=[]; let prev=null;
  const sentences=[]; let cur=[];
  for(const wd of words){ if(wd.sentenceEnd){sentences.push(cur);cur=[];} else cur.push(wd); }
  if(cur.length)sentences.push(cur);
  for(const sent of sentences){
    let total=0;
    for(const wd of sent){ wd.segs=segments(wd.ph); wd.dur=wd.segs.reduce((a,s)=>a+s.dur,0); total+=wd.dur+(wd.pause||0); }
    let t=0; const f0a=118*pitch, f0b=84*pitch;
    for(let wi=0;wi<sent.length;wi++){
      const wd=sent[wi]; const finalW=wi===sent.length-1;
      if(finalW){ for(let i=wd.segs.length-1;i>=0;i--){ if(['vowel','nasal','liquid'].includes(wd.segs[i].kind)){wd.segs[i].dur=wd.segs[i].dur*1.55|0;break;} } wd.dur=wd.segs.reduce((a,s)=>a+s.dur,0); }
      const b0=f0a+(f0b-f0a)*(t/Math.max(1,total)); t+=wd.dur; const b1=f0a+(f0b-f0a)*(t/Math.max(1,total)); t+=(wd.pause||0);
      let acc=0;
      for(const seg of wd.segs){
        const dur=seg.dur/rate, tr=Math.min(seg.tr,dur*.6);
        const nf=Math.max(1,Math.round(dur/5)), nt=Math.max(0,Math.round(tr/5));
        let tp=seg.t;
        if(seg.kind==='vowel')tp=balance(seg.t,'vowel');
        else if(['nasal','liquid','glide','fric','burst','aspir'].includes(seg.kind))tp=balance(seg.t,seg.kind);
        else if(seg.kind==='sil'&&seg.t.av)tp=balance(seg.t,'voicebar');
        for(let k=0;k<nf;k++){
          const fr=Object.assign({},tp);
          if(k<nt&&prev){const mix=(k+1)/(nt+1);for(const key of new Set([...Object.keys(prev),...Object.keys(tp)]))fr[key]=(prev[key]||0)+((tp[key]||0)-(prev[key]||0))*mix;}
          const frac=(acc/5)/Math.max(1,wd.dur);
          let f=b0+(b1-b0)*frac;
          if(seg.stress===1&&seg.kind==='vowel')f*=1+.14*Math.sin(Math.PI*Math.min(1,frac*1.4));
          if(finalW&&seg.kind==='vowel')f*=1-.13*frac*frac;
          fr.f0=f; frames.push(fr); acc+=5;
        }
        prev=tp;
      }
      if(wd.pause){for(let k=0;k<Math.round(wd.pause/rate/5);k++)frames.push({f0:95*pitch});prev=null;}
    }
  }
  return frames;
}
function renderPCM(text, rate, pitch){
  const pcm = synthesize(framesForText(text, rate, pitch));
  let peak=1e-6; for(const x of pcm)peak=Math.max(peak,Math.abs(x));
  const g=.89/peak, nf=80;
  for(let i=0;i<pcm.length;i++)pcm[i]*=g;
  for(let i=0;i<Math.min(nf,pcm.length);i++){pcm[i]*=i/nf;pcm[pcm.length-1-i]*=i/nf;}
  return pcm;
  }

File: tts.py
"""
tts.py - text -> klatt80 parameter frames -> PCM.

Front end: CMUdict (BSD 2-clause, (c) Carnegie Mellon University) for
pronunciation, simple OOV letter rules, digit expansion, and a small
declination/accent prosody model in the 1980 formant-synth style.
MIT license - Humanico Voice project.
"""
import re, math, wave, struct, sys, os
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import klatt80
from phones import (VOWELS, DIPHTHONGS, VOWEL_SET, AV_VOWEL, STOPS, FRICATIVES,
                    NASALS, LIQUIDS, GLIDES, AFFRICATES)

FS = klatt80.FS
FRAME_MS = klatt80.FRAME_MS

# ---------------------------------------------------------------- dictionary
def load_cmudict(path):
    d = {}
    with open(path, encoding='latin-1') as fh:
        for line in fh:
            line = line.strip()
            if not line or line.startswith(';;;'):
                continue
            parts = line.split()
            word = re.sub(r'\(\d+\)$', '', parts[0]).lower()
            if word not in d:
                d[word] = parts[1:]
    return d

_DICT = None
def cmu():
    global _DICT
    if _DICT is None:
        _DICT = load_cmudict(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'cmudict.dict'))
    return _DICT

DIGITS = {'0':'zero','1':'one','2':'two','3':'three','4':'four','5':'five',
          '6':'six','7':'seven','8':'eight','9':'nine'}

# crude letter-to-sound fallback for out-of-vocabulary words
LTS = {'a':'AH','b':'B','c':'K','d':'D','e':'EH','f':'F','g':'G','h':'HH',
       'i':'IH','j':'JH','k':'K','l':'L','m':'M','n':'N','o':'AA','p':'P',
       'q':'K','r':'R','s':'S','t':'T','u':'AH','v':'V','w':'W','x':'K S',
       'y':'Y','z':'Z'}

def word_to_phones(word):
    """-> list of (phone, stress) where stress in {0,1,2}; vowels only."""
    w = word.lower()
    if w.isdigit():
        out = []
        for ch in w:
            out += word_to_phones(DIGITS[ch])
        return out
    entry = cmu().get(w)
    if entry is None:
        entry = []
        for ch in w:
            if ch in LTS:
                entry += LTS[ch].split()
        if not entry:
            return []
    out = []
    for p in entry:
        m = re.match(r'^([A-Z]+)([012])?$', p)
        ph, stress = m.group(1), int(m.group(2) or 0)
        if ph == 'AXR':
            ph, stress = 'ER', stress
        if ph in VOWEL_SET:
            out.append((ph, stress))
        else:
            out.append((ph, -1))
    return out


# ------------------------------------------------------------- auto-calibration
# Balance source amplitudes so every segment lands at a target loudness
# relative to a reference vowel, whatever the resonator gains do.
_RMS_CACHE = {}

def _unit_rms(params, source):
    """Steady-state RMS with `source` amplitude = 1, others 0."""
    key = (source, tuple(sorted((k, round(v, 1)) for k, v in params.items()
                                if k != source and isinstance(v, (int, float)))))
    if key in _RMS_CACHE:
        return _RMS_CACHE[key]
    v = klatt80.Voice()
    for k, val in params.items():
        setattr(v, k, val)
    for s in ('av', 'ah', 'af'):
        setattr(v, s, 0.0)
    setattr(v, source, 1.0)
    if not params.get('f0'):
        v.f0 = 110.0
    pcm = []
    for _ in range(70):
        pcm.extend(v.frame(klatt80.FRAME))
    import numpy as _np
    x = _np.array(pcm[1500:])
    rms = float(_np.sqrt((x ** 2).mean())) + 1e-9
    _RMS_CACHE[key] = rms
    return rms

_VREF = {}

def _vowel_ref():
    if 'r' not in _VREF:
        _VREF['r'] = _unit_rms(dict(f1=730.0, b1=85.0, f2=1090.0, b2=110.0,
                                    f3=2440.0, b3=170.0), 'av')
    return _VREF['r']

# desired levels relative to reference vowel (dB)
LEVEL = dict(vowel=0.0, nasal=-4.0, liquid=-4.0, glide=-2.0,
             fric=-7.0, fric_weak=-10.0, burst=-5.0, aspir=-12.0,
             voicebar=-20.0, fric_voice=-16.0)

def _set_level(target, source, kind):
    db = LEVEL[kind]
    desired = _vowel_ref() * (10 ** (db / 20.0))
    unit = _unit_rms(target, source)
    target[source] = desired / unit

def balance(target, kind):
    t = dict(target)
    if kind == 'vowel':
        _set_level(t, 'av', 'vowel')
    elif kind in ('nasal', 'liquid'):
        _set_level(t, 'av', kind)
    elif kind == 'glide':
        _set_level(t, 'av', 'glide')
    elif kind == 'fric':
        _set_level(t, 'af', 'fric_weak' if t.get('ab') and not (t.get('a4') or t.get('a5') or t.get('a6')) else 'fric')
        if t.get('av'):
            _set_level(t, 'av', 'fric_voice')
    elif kind == 'burst':
        _set_level(t, 'af', 'burst')
    elif kind == 'aspir':
        _set_level(t, 'ah', 'aspir')
    elif kind == 'voicebar':
        _set_level(t, 'av', 'voicebar')
    return t

# ------------------------------------------------------------------- segments
SIL = dict(av=0.0, ah=0.0, af=0.0, ab=0.0, a2=0.0, a3=0.0, a4=0.0, a5=0.0, a6=0.0)

def tgt(**kw):
    t = dict(SIL); t.update(kw); return t

def vowel_tgt(v, av=AV_VOWEL):
    f1, f2, f3, b1, b2, b3 = VOWELS[v]
    return tgt(av=av, f1=f1, f2=f2, f3=f3, b1=b1, b2=b2, b3=b3)

def phones_to_segments(phseq):
    """phseq: list of (phone, stress). -> list of dicts:
       {target: {...}, dur_ms, trans_ms, kind, stress}"""
    segs = []
    n = len(phseq)
    for i, (ph, stress) in enumerate(phseq):
        nxt = phseq[i + 1][0] if i + 1 < n else None
        stressed = stress == 1
        if ph in VOWELS:
            dur = 150 if stressed else (100 if stress == 2 else 85)
            segs.append(dict(target=balance(vowel_tgt(ph), 'vowel'), dur_ms=dur, trans_ms=50,
                             kind='vowel', stress=stress))
        elif ph in DIPHTHONGS:
            a, b = DIPHTHONGS[ph]
            dur = 165 if stressed else 115
            t1 = balance(vowel_tgt(a), 'vowel'); t2 = balance(vowel_tgt(b), 'vowel')
            segs.append(dict(target=t1, dur_ms=int(dur * 0.45), trans_ms=55,
                             kind='vowel', stress=stress))
            segs.append(dict(target=t2, dur_ms=int(dur * 0.55), trans_ms=int(dur * 0.5),
                             kind='vowel', stress=stress))
        elif ph in STOPS:
            sp = STOPS[ph]
            closure = dict(SIL)
            if sp['voiced']:
                closure.update(av=0.18, f1=200.0, b1=250.0, f2=1000.0, f3=2200.0)
            bt = 'voicebar' if sp['voiced'] else None
            if bt:
                closure = balance(tgt(**closure), 'voicebar')
            segs.append(dict(target=tgt(**closure), dur_ms=sp['closure'], trans_ms=3,
                             kind='sil', stress=-1))
            bp = dict(af=sp['burst'].get('af', 0.6))
            for k in ('a2', 'a3', 'a4', 'a5', 'a6', 'ab'):
                if k in sp['burst']:
                    bp[k] = sp['burst'][k]
            segs.append(dict(target=balance(tgt(**bp), 'burst'), dur_ms=14, trans_ms=3,
                             kind='burst', stress=-1))
            if sp['aspir'] > 0:
                ah = dict(ah=sp['ah'])
                if nxt in VOWEL_SET:
                    v = vowel_tgt(DIPHTHONGS.get(nxt, (nxt,))[0] if nxt in DIPHTHONGS else nxt)
                    ah.update(f1=v['f1'], f2=v['f2'], f3=v['f3'])
                segs.append(dict(target=balance(tgt(**ah), 'aspir'), dur_ms=sp['aspir'], trans_ms=12,
                                 kind='aspir', stress=-1))
        elif ph in AFFRICATES:
            st, fr = AFFRICATES[ph]
            sub = phones_to_segments([(st, -1), (fr, -1)])
            sub[0]['dur_ms'] = int(sub[0]['dur_ms'] * 0.7)
            sub = [s for s in sub if s['kind'] != 'aspir']
            segs += sub
        elif ph in FRICATIVES:
            dur, prm = FRICATIVES[ph]
            t = tgt(**{k: v for k, v in prm.items()})
            if ph == 'HH' and nxt in VOWEL_SET:
                v = vowel_tgt(DIPHTHONGS.get(nxt, (nxt,))[0] if nxt in DIPHTHONGS else nxt)
                t.update(f1=v['f1'], f2=v['f2'], f3=v['f3'])
            segs.append(dict(target=balance(t, 'fric'), dur_ms=dur, trans_ms=20, kind='fric', stress=-1))
        elif ph in NASALS:
            dur, f2, fnz = NASALS[ph]
            segs.append(dict(target=balance(tgt(av=0.75, f1=270.0, b1=250.0, f2=f2, b2=200.0,
                                        f3=2400.0, b3=250.0, fnz=fnz, bnz=250.0), 'nasal'),
                             dur_ms=dur, trans_ms=35, kind='nasal', stress=-1))
        elif ph in LIQUIDS:
            dur, f1, f2, f3, b1, b2 = LIQUIDS[ph]
            segs.append(dict(target=balance(tgt(av=0.72, f1=f1, f2=f2, f3=f3, b1=b1, b2=b2, b3=200.0), 'liquid'),
                             dur_ms=dur, trans_ms=45, kind='liquid', stress=-1))
        elif ph in GLIDES:
            dur, f1, f2, f3 = GLIDES[ph]
            segs.append(dict(target=balance(tgt(av=0.80, f1=f1, f2=f2, f3=f3, b1=90.0, b2=130.0, b3=180.0), 'glide'),
                             dur_ms=dur, trans_ms=45, kind='glide', stress=-1))
    return segs

# ------------------------------------------------------------------- prosody
PAUSE = {',': 200, ';': 220, ':': 220, '-': 160, '.': 340, '!': 340, '?': 340}

def tokenize(text):
    text = text.replace('\u2014', ' - ').replace('\u2013', ' - ').replace('\u2026', '...')
    raw = re.findall(r"[A-Za-z0-9']+|[.,!?;:\-]", text)
    out = []
    for r in raw:
        if r in '.,!?;:-':
            out.append(('punct', r))
        else:
            out.append(('word', r.strip("'")))
    return out

def f0_track(sent_words, total_ms):
    """Declination contour with accent bumps; sent_words = list of word dicts."""
    f0 = []
    for wd in sent_words:
        base0, base1 = wd['f0a'], wd['f0b']
        for seg in wd['segs']:
            n = max(1, int(seg['dur_ms'] / FRAME_MS))
            for k in range(n):
                frac = k / max(1, n - 1)
                f = base0 + (base1 - base0) * frac
                if seg.get('stress') == 1 and seg['kind'] == 'vowel':
                    bump = math.sin(math.pi * min(1.0, frac * 1.4))
                    f *= 1.0 + 0.14 * bump
                if wd.get('final') and seg['kind'] == 'vowel':
                    f *= 1.0 - 0.13 * frac * frac
                f0.append(f)
    return f0

def utterance_frames(text, rate=1.0, pitch=1.0):
    """text -> (frames, total_ms). Frames carry klatt80 params incl. f0."""
    tokens = tokenize(text)
    sentences, cur = [], []
    for kind, val in tokens:
        if kind == 'punct':
            if cur:
                cur[-1]['pause_after'] = PAUSE.get(val, 200)
                if val in '.!?':
                    sentences.append(cur); cur = []
                else:
                    cur[-1]['comma'] = True
        elif val:
            ph = word_to_phones(val)
            if ph:
                cur.append(dict(word=val, phones=ph, pause_after=0, comma=False))
    if cur:
        sentences.append(cur)

    frames = []
    for sent in sentences:
        # collect segments per word first (for durations + total length)
        total_ms = 0
        for wi, wd in enumerate(sent):
            segs = phones_to_segments(wd['phones'])
            # phrase-final lengthening
            if wi == len(sent) - 1:
                for s in reversed(segs):
                    if s['kind'] in ('vowel', 'nasal', 'liquid'):
                        s['dur_ms'] = int(s['dur_ms'] * 1.55)
                        break
                wd['final'] = True
            wd['segs'] = segs
            wd['dur_ms'] = sum(s['dur_ms'] for s in segs)
            total_ms += wd['dur_ms'] + wd['pause_after']
        # declination endpoints
        f_start, f_end = 118.0 * pitch, 84.0 * pitch
        t = 0.0
        for wd in sent:
            a = f_start + (f_end - f_start) * (t / max(1, total_ms))
            t += wd['dur_ms']
            b = f_start + (f_end - f_start) * (t / max(1, total_ms))
            wd['f0a'], wd['f0b'] = a, b
            t += wd['pause_after']

        # render segments + f0
        f0s = f0_track(sent, total_ms)
        fi = 0
        prev = None
        for wd in sent:
            for seg in wd['segs']:
                dur = seg['dur_ms'] / rate
                trans = min(seg['trans_ms'], dur * 0.6)
                n_frames = max(1, round(dur / FRAME_MS))
                n_trans = max(0, round(trans / FRAME_MS))
                tgt_params = seg['target']
                if prev is None:
                    prev = tgt_params
                for k in range(n_frames):
                    fr = dict(tgt_params)
                    if k < n_trans and prev is not tgt_params:
                        mix = (k + 1) / (n_trans + 1)
                        for key in set(prev) | set(tgt_params):
                            p0 = prev.get(key, 0.0); p1 = tgt_params.get(key, 0.0)
                            fr[key] = p0 + (p1 - p0) * mix
                    if fi < len(f0s):
                        fr['f0'] = f0s[fi]
                    fi += 1
                    frames.append(fr)
                prev = tgt_params
            # word pause
            if wd['pause_after']:
                for _ in range(round(wd['pause_after'] / rate / FRAME_MS)):
                    frames.append(dict(SIL, f0=95.0 * pitch))
                prev = None
    return frames

# ------------------------------------------------------------------- rendering
def render(text, rate=1.0, pitch=1.0):
    frames = utterance_frames(text, rate=rate, pitch=pitch)
    pcm = klatt80.synthesize(frames)
    nf = int(0.008 * FS)
    for i in range(min(nf, len(pcm))):
        pcm[i] *= i / nf
        pcm[-1 - i] *= i / nf
    peak = max(1e-6, max(abs(x) for x in pcm))
    g = 0.89 / peak
    return [max(-1.0, min(1.0, x * g)) for x in pcm]

def write_wav(path, pcm, fs=FS):
    with wave.open(path, 'wb') as w:
        w.setnchannels(1); w.setsampwidth(2); w.setframerate(fs)
        w.writeframes(b''.join(struct.pack('<h', int(x * 32767)) for x in pcm))

if __name__ == '__main__':
    text = ' '.join(sys.argv[2:]) if len(sys.argv) > 2 else 'It begins with one of us.'
    pcm = render(text)
    write_wav(sys.argv[1], pcm)
    print(f'{sys.argv[1]}: {len(pcm)/FS:.2f}s')
